File: Not_Today_Sun/app/src/main/java/com/example/not_today_sun/settings/viewmodel/SettingsViewModel.java
package com.example.not_today_sun.settings.viewmodel;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import androidx.lifecycle.ViewModel;

public class SettingsViewModel extends ViewModel {
    private static final String TAG = "SettingsViewModel";
    private static final String PREFS_NAME = "WeatherSettings";
    private static final String KEY_LOCATION = "location";
    private static final String KEY_TEMPERATURE = "temperature";
    private static final String KEY_LANGUAGE = "language";
    private static final String KEY_WIND_SPEED = "wind_speed";
    private static final String KEY_NOTIFICATIONS = "notifications";

    public void saveSettings(Context context, String location, String temperature,
                             String language, String windSpeed, boolean notifications) {
        SharedPreferences sharedPref = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();

        Log.d(TAG, "Saving settings to SharedPreferences:");
        Log.d(TAG, "Location: " + location);
        Log.d(TAG, "Temperature: " + temperature);
        Log.d(TAG, "Language: " + language);
        Log.d(TAG, "Wind Speed: " + windSpeed);
        Log.d(TAG, "Notifications: " + notifications);

        editor.putString(KEY_LOCATION, location);
        editor.putString(KEY_TEMPERATURE, temperature);
        editor.putString(KEY_LANGUAGE, language);
        editor.putString(KEY_WIND_SPEED, windSpeed);
        editor.putBoolean(KEY_NOTIFICATIONS, notifications);
        editor.apply();

        Log.d(TAG, "Settings successfully saved to SharedPreferences");
    }

    public String getLocationSetting(Context context) {
        SharedPreferences sharedPref = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String value = sharedPref.getString(KEY_LOCATION, "gps");
        Log.d(TAG, "Getting location setting: " + value);
        return value;
    }

    public String getTemperatureSetting(Context context) {
        SharedPreferences sharedPref = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String value = sharedPref.getString(KEY_TEMPERATURE, "celsius");
        Log.d(TAG, "Getting temperature setting: " + value);
        return value;
    }

    public String getLanguageSetting(Context context) {
        SharedPreferences sharedPref = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String value = sharedPref.getString(KEY_LANGUAGE, "english");
        Log.d(TAG, "Getting language setting: " + value);
        return value;
    }

    public String getWindSpeedSetting(Context context) {
        SharedPreferences sharedPref = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String value = sharedPref.getString(KEY_WIND_SPEED, "meter_sec");
        Log.d(TAG, "Getting wind speed setting: " + value);
        return value;
    }

    public boolean getNotificationsSetting(Context context) {
        SharedPreferences sharedPref = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        boolean value = sharedPref.getBoolean(KEY_NOTIFICATIONS, true);
        Log.d(TAG, "Getting notifications setting: " + value);
        return value;
    }
}

File: Not_Today_Sun/app/src/main/java/com/example/not_today_sun/settings/viewmodel/SettingsViewModel.kt
package com.example.not_today_sun.settings.viewmodel

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import java.util.Locale

class SettingsViewModel : ViewModel() {
    companion object {
        private const val PREFS_NAME = "WeatherSettings"
        private const val KEY_LOCATION = "use_gps"
        private const val KEY_MAP = "use_map"
        private const val KEY_NOTIFICATIONS = "notifications_enabled"
        private const val KEY_MAP_LAT = "map_lat"
        private const val KEY_MAP_LON = "map_lon"
        private const val KEY_MAP_LOCATION_NAME = "map_location_name"
        private const val KEY_TEMPERATURE_UNIT = "temperature_unit"
        private const val KEY_WIND_SPEED_UNIT = "wind_speed_unit"
        private const val KEY_LANGUAGE = "language"
    }

    private val _notificationEnabled = MutableLiveData<Boolean>()

    init {
        _notificationEnabled.value = false
    }

    private fun getSharedPreferences(context: Context): SharedPreferences {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    fun getLocationPreference(context: Context): Boolean {
        return getSharedPreferences(context).getBoolean(KEY_LOCATION, true)
    }

    fun getMapPreference(context: Context): Boolean {
        return getSharedPreferences(context).getBoolean(KEY_MAP, false)
    }

    fun getNotificationPreference(context: Context): Boolean {
        val enabled = getSharedPreferences(context).getBoolean(KEY_NOTIFICATIONS, false)
        _notificationEnabled.postValue(enabled)
        return enabled
    }

    fun saveLocationPreference(context: Context, useGps: Boolean, useMap: Boolean) {
        with(getSharedPreferences(context).edit()) {
            putBoolean(KEY_LOCATION, useGps)
            putBoolean(KEY_MAP, useMap)
            if (!useMap) {
                remove(KEY_MAP_LAT)
                remove(KEY_MAP_LON)
                remove(KEY_MAP_LOCATION_NAME)
            }
            apply()
        }
    }

    fun saveMapLocation(context: Context, latitude: Float, longitude: Float, locationName: String) {
        with(getSharedPreferences(context).edit()) {
            putFloat(KEY_MAP_LAT, latitude)
            putFloat(KEY_MAP_LON, longitude)
            putString(KEY_MAP_LOCATION_NAME, locationName)
            putBoolean(KEY_LOCATION, false)
            putBoolean(KEY_MAP, true)
            apply()
        }
    }

    fun saveNotificationPreference(context: Context, enabled: Boolean) {
        with(getSharedPreferences(context).edit()) {
            putBoolean(KEY_NOTIFICATIONS, enabled)
            apply()
        }
        _notificationEnabled.postValue(enabled)
    }

    fun saveTemperatureUnit(context: Context, unit: String) {
        with(getSharedPreferences(context).edit()) {
            putString(KEY_TEMPERATURE_UNIT, unit)
            apply()
        }
    }

    fun getTemperatureUnit(context: Context): String {
        return getSharedPreferences(context).getString(KEY_TEMPERATURE_UNIT, "standard") ?: "standard"
    }

    fun saveWindSpeedUnit(context: Context, unit: String) {
        with(getSharedPreferences(context).edit()) {
            putString(KEY_WIND_SPEED_UNIT, unit)
            apply()
        }
    }

    fun getWindSpeedUnit(context: Context): String {
        return getSharedPreferences(context).getString(KEY_WIND_SPEED_UNIT, "m/s") ?: "m/s"
    }

    fun saveLanguage(context: Context, language: String) {
        with(getSharedPreferences(context).edit()) {
            putString(KEY_LANGUAGE, language)
            apply()
        }
        // Apply locale change
        updateLocale(context, language)
    }

    private fun updateLocale(context: Context, languageCode: String) {
        val locale = Locale(languageCode)
        Locale.setDefault(locale)
        val config = Configuration(context.resources.configuration)
        config.setLocale(locale)
        context.resources.updateConfiguration(config, context.resources.displayMetrics)
    }

    fun getLanguage(context: Context): String {
        return getSharedPreferences(context).getString(KEY_LANGUAGE, "en") ?: "en"
    }
}